src/tools/hashing.rs

```diff
@@ -46,27 +48,30 @@ impl Chain for XofHasher {
     }
 }
 
-impl XofHasher {
+impl<P: SchemeParams> Hasher<P> {
     fn new() -> Self {
-        Self(Shake256::default())
+        Self(P::Digest::default())
     }
 
     pub fn new_with_dst(dst: &[u8]) -> Self {
         Self::new().chain_bytes(dst)
     }
 
-    pub fn finalize_to_reader(self) -> Shake256Reader {
+    pub fn finalize_to_reader(self) -> <P::Digest as ExtendableOutput>::Reader {
         self.0.finalize_xof()
     }
 
-    /// Finalizes into enough bytes to bring the collision probability under `2^(-security_bits)`.
-    pub fn finalize_boxed(self, security_bits: usize) -> Box<[u8]> {
+    /// Finalizes into enough bytes to bring the collision probability to what's required by the scheme's security.
+    pub fn finalize(self) -> HashOutput {
         // A common heuristic for hashes is that the log2 of the collision probability is half the output size.
         // We may not have enough output bytes, but this constitutes the best effort.
-        self.0.finalize_xof().read_boxed((security_bits * 2).div_ceil(8))
+        HashOutput(self.0.finalize_xof().read_boxed((P::SECURITY_BITS * 2).div_ceil(8)))
     }
 }
 
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub struct HashOutput(#[serde(with = "SliceLike::<Hex>")] Box<[u8]>);
```

Comment:
I think that the previous code being decoupled from `SchemeParams` was a good thing and that the return of `HashOutput` is a "downgrade".

Reply:
The wrapper is there mainly to encapsulate the serialization attribute. 

As for the hasher being a part of scheme parameters, while it adds to the complexity somewhat, I think it is a logical  and, I guess, inevitable consequence of `SECURITY_BITS` being a user-defined parameter.

Reply:
Your intuition was right, being generic over parameters didn't really work out, which I realized when I added hashing all the scheme parameters. Hashing Paillier parameters using the scheme as the generic would be weird.